test/perf/perf_alternator_workloads.cc

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (C) 2023-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include <functional>
+#include <memory>
+#include <signal.h>
+#include <seastar/core/thread.hh>
+#include <seastar/core/app-template.hh>
+#include <seastar/http/client.hh>
+#include <seastar/http/request.hh>
+#include <seastar/http/reply.hh>
+#include <seastar/util/defer.hh>
+#include <seastar/util/short_streams.hh>
+#include <tuple>
+#include <boost/program_options.hpp>
+
+#include "db/config.hh"
+#include "seastar/core/future.hh"
+#include "test/perf/perf.hh"
+#include "test/lib/random_utils.hh"
+
+
+namespace perf {
+
+using namespace seastar;
+namespace bpo = boost::program_options;
+
+struct test_config {
+    std::string workload;
+    int port;
+    unsigned partitions;
+    unsigned duration_in_seconds;
+    unsigned concurrency;
+    bool flush;
+};
+
+std::ostream& operator<<(std::ostream& os, const test_config& cfg) {
+    return os << "{workload=" << cfg.workload
+           << ", partitions=" << cfg.partitions
+           << ", concurrency=" << cfg.concurrency
+           << ", duration_in_seconds=" << cfg.duration_in_seconds
+           << ", flush=" << cfg.flush
+           << "}";
+}
+
+static http::experimental::client get_client(int port) {
+    return http::experimental::client(socket_address(net::inet_address("127.0.0.1"), port));
+}
+
+static future<> make_request(http::experimental::client& cli, sstring operation, sstring body) {
+    auto req = http::request::make("POST", "localhost", "/");
+    req._headers["X-Amz-Target:"] = "DynamoDB_20120810." + operation;
+    req.write_body("application/x-amz-json-1.0", std::move(body));
+    co_await cli.make_request(std::move(req), [] (const http::reply& rep, input_stream<char>&& in_) -> future<> {
+        auto in = std::move(in_);
```

Comment:
It's probably to avoid the lambda coroutine fiasco. Please see if coroutine::lambda() fixes it instead.

Reply:
I thought about this, but wasn't this fiasco about captures, not parameters? (but maybe I'm misremembering).

Reply:
yes, it was added to keep it alive. Indeed coroutine fiasco document mentions only captures. I am not sure why this is happening. It looks like `in_` is freed after the first suspension point in lambda.

Adding coroutine::lambda or the second solution with std::ref doesn't help here.

Perhaps there is something wrong with caller code in seastar?
```c++
            return do_with(std::move(rep), [&con, handle = std::move(handle)] (auto& rep) mutable {
                return handle(rep, con.in(rep));
            });
```
`handle` is my lambda from above. I've tried also put it in do_with to keep it alive, without success:
```c++
            return do_with(std::move(rep), std::move(handle), [&con] (auto& rep, auto& handle) mutable {
                return handle(rep, con.in(rep));
            });
```

so this trick with 
`auto in = std::move(in_);` is the only one which works for me (I saw it also here: https://github.com/scylladb/scylladb/blob/master/alternator/executor.cc#L91 and in one of Pavel's wip patches)

Reply:
like just wrap the lambda? 
```c++
co_await cli.make_request(std::move(req), coroutine::lambda([] (const http::reply& rep, input_stream<char>&& in_) -> future<> {
...
})
```
tested this before and it doesn't solve it

Reply:
I think it can be solved but in seastar instead:

```diff
diff --git a/src/http/client.cc b/src/http/client.cc
index 7a66823c..46df99df 100644
--- a/src/http/client.cc
+++ b/src/http/client.cc
@@ -236,9 +236,9 @@ future<> client::make_request(request req, reply_handler handle, reply::status_t
                 return make_exception_future<>(std::runtime_error(format("request finished with {}", rep._status)));
             }
 
-            return do_with(std::move(rep), [&con, handle = std::move(handle)] (auto& rep) mutable {
-                return handle(rep, con.in(rep));
-            });
+            return do_with(std::move(rep), coroutine::lambda([&con, handle = std::move(handle)] (auto& rep) mutable -> future<> {
+                co_await handle(rep, con.in(rep));
+            }));
         });
     });
 }
```


while this is documented 
`
/// lambda coroutine must complete (co_await) in the same statement.
`
I am not sure why it requires immediate co_await to work